src/core/provider.h

```diff
@@ -35,6 +35,42 @@ namespace nvidia { namespace inferenceserver {
 
 class InferenceBackend;
 
+//
+// Memory block used for accessing / storing data for providers
+// Note:
+//   - It is only responsible for deallocating memory obtained by ReserveBlock()
+//   - The functions are not thread-safe
+//
+class SystemMemoryBlock {
+ public:
+  // Create an SystemMemoryBlock object with block size 0
+  SystemMemoryBlock();
+
+  // Create an SystemMemoryBlock that points to 'block' provided
+  // 'block' is the memory block that has been allocated. For instance,
+  // content returned by InferRequestProvider.GetNextInputContent().
+  // 'block_size' is the size of the block
+  SystemMemoryBlock(void* block, size_t block_size);
+
+  // Reserve a block of memory with size 'byte_size'
+  // 'byte_size' is the size of the memory to be reserved
+  // Return error if the SystemMemoryBlock has reserved memory
+  // with non-zero size
+  Status ReserveBlock(size_t byte_size);
+
+  // Return a pointer to the memory location. It is caller's responsibility
+  // to make sure the access to the memory is exclusive on write.
+  void* GetBlock() const;
```

Comment:
We don't typically use "Get" prefix (and note you don't say GetByteSize). So I would change to just "Block".